File: dynamic-programming/dayso/dayso.py
import os
import sys

input_file = os.path.join(sys.path[0], 'dayso1.inp')
output_file = os.path.join(sys.path[0], 'dayso1.out')

INF = float('inf')

n = K = 0
a = []
result = 0

# d[i] là điểm số lớn nhất khi người chơi dừng cuộc chơi tại vị trí i
d = []

def input():
    global n, K, a

    sys.stdin = open(input_file, 'r')
    data = sys.stdin.read().split()
    iterator = iter(data)

    n = int(next(iterator))
    K = int(next(iterator))
    
    a = [0] * (n + 1)
    
    for i in range(1, n + 1):
        x = int(next(iterator))
        a[i] = x
    

def process():
    global result

    # Khởi tạo mảng d
    d = [-INF] * (n + 1)
    d[0] = 0

    # Duyệt từng ô trong dãy số
    for i in range(1, n + 1):
        # Duyệt số bước nhảy
        for k in range(1, K + 1):
            if i - k >= 0:
                d[i] = max(d[i], d[i - k] + a[i])

        # Cập nhật điểm số tổng thể lớn nhất
        result = max(result, d[i])
    

def output():
    sys.stdout = open(output_file, 'w')
    
    print(result)


if __name__ == '__main__':
    input()
    process()
    output()

File: dynamic-programming/dayso/dayso.cpp
#include <iostream>
#include <vector>
#include <algorithm>

#define input_file "dayso1.inp"
#define output_file "dayso1.out"

using namespace std;

const int INF = 1e9;

int n, K;
vector<int> a;
int result;

// d[i] là điểm số lớn nhất khi người chơi dừng cuộc chơi tại vị trí i
vector<int> d;

void input()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    freopen(input_file, "r", stdin);

    cin >> n >> K;

    a.resize(n + 1, 0);

    for (int i = 1; i <= n; ++i)
    {
        cin >> a[i];
    }
}

void process()
{
    // Khởi tạo mảng d
    d.resize(n + 1, -INF);
    d[0] = 0;

    // Duyệt từng ô trong dãy số
    for (int i = 1; i <= n; ++i)
    {
        // Duyệt số bước nhảy
        for (int k = 1; k <= K; ++k)
        {
            if (i - k >= 0)
            {
                d[i] = max(d[i], d[i - k] + a[i]);
            }
        }

        // Cập nhật điểm số tổng thể lớn nhất
        result = max(result, d[i]);
    }
}

void output()
{
    freopen(output_file, "w", stdout);

    cout << result;
}

int main()
{
    input();
    process();
    output();

    return 0;
}